src/GRANULAR/granular_model.cpp

```diff
@@ -0,0 +1,493 @@
+    fwrite(sub_models[i]->name.data(), sizeof(char), num_char, fp);
+    fwrite(&num_coeffs, sizeof(int), 1, fp);
+    fwrite(sub_models[i]->coeffs, sizeof(double), num_coeffs, fp);
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+void GranularModel::read_restart(FILE *fp)
+{
+  int num_char, num_coeff;
+
+  for (int i = 0; i < NSUBMODELS; i++) {
+    if (comm->me == 0)
+      utils::sfread(FLERR, &num_char, sizeof(int), 1, fp, nullptr, error);
+    MPI_Bcast(&num_char, 1, MPI_INT, 0, world);
+
+    std::string model_name (num_char, ' ');
+    if (comm->me == 0)
+      utils::sfread(FLERR, const_cast<char*>(model_name.data()), sizeof(char),num_char, fp, nullptr, error);
+    MPI_Bcast(const_cast<char*>(model_name.data()), num_char, MPI_CHAR, 0, world);
+
+    construct_submodel(model_name, (SubmodelType) i);
+
+    if (comm->me == 0) {
+      utils::sfread(FLERR, &num_coeff, sizeof(int), 1, fp, nullptr, error);
+      if (num_coeff != sub_models[i]->num_coeffs)
+        error->one(FLERR, "Invalid granular model written to restart file");
+    }
+    MPI_Bcast(&num_coeff, 1, MPI_INT, 0, world);
+
+    if (comm->me == 0) {
+      utils::sfread(FLERR, sub_models[i]->coeffs, sizeof(double), num_coeff, fp, nullptr, error);
+    }
+    MPI_Bcast(sub_models[i]->coeffs, num_coeff, MPI_DOUBLE, 0, world);
+    sub_models[i]->coeffs_to_local();
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+bool GranularModel::check_contact()
+{
+  if (contact_type == WALL) {
+    // Used by fix_wall_gran.cpp
+    //   radj = radius of wall
+    //   dx already provided
+    rsq = lensq3(dx);
+    radsum = radi;
+    if (radj == 0) Reff = radi;
+    else Reff = radi * radj / (radi + radj);
+  } else if (contact_type == WALLREGION) {
+    // Used by fix_wall_gran_region.cpp
+    //   radj = radius of wall
+    //   dx and r already provided
+    rsq = r * r;
+    radsum = radi;
+    if (radj == 0) Reff = radi;
+    else Reff = radi * radj / (radi + radj);
+  } else {
+    sub3(xi, xj, dx);
+    rsq = lensq3(dx);
+    radsum = radi + radj;
+    Reff = radi * radj / radsum;
+  }
+
+  touch = normal_model->touch();
+  return touch;
+}
+
+/* ---------------------------------------------------------------------- */
+
+void GranularModel::calculate_forces()
+{
+  // Standard geometric quantities
+
+  if (contact_type != WALLREGION) r = sqrt(rsq);
+  rinv = 1.0 / r;
+  delta = radsum - r;
+  dR = delta * Reff;
+  scale3(rinv, dx, nx);
+
+  // relative translational velocity
+  sub3(vi, vj, vr);
+
+  // normal component
+  vnnr = dot3(vr, nx); //v_R . n
+  scale3(vnnr, nx, vn);
+
+  // tangential component
+  sub3(vr, vn, vt);
+
+  // relative rotational velocity
+  scaleadd3(radi, omegai, radj, omegaj, wr);
+
+  // relative tangential velocities
+  double temp[3];
+  cross3(wr, nx, temp);
+  sub3(vt, temp, vtr);
+  vrel = len3(vtr);
+
+  // calculate forces/torques
+  forces[0] = 0.0;
```

Comment:
Is this supposed to clear out the entire "forces" array? Or just the first element?

Reply:
No, that looks like an overlooked line that should have been deleted earlier. Removed now.